File: 0095-unique-binary-search-trees-ii/0095-unique-binary-search-trees-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> generateTrees(int n) {
        return helper(1, n);
    }
    
private:
    vector<TreeNode*> helper(int left, int right){
        vector<TreeNode*> result;
        if(left > right){
            result.push_back(nullptr);
            return result;
        }
        
        for(int node=left;node<=right;node++){
            vector<TreeNode*> leftNodes = helper(left, node-1);
            vector<TreeNode*> rightNodes = helper(node+1, right);
            
            for(TreeNode *l:leftNodes){
                for(TreeNode *r:rightNodes){
                    TreeNode *root = new TreeNode(node);
                    root->left = l;
                    root->right = r;
                    result.push_back(root);
                }
            }
        }
        
        return result;
    }
};

File: 0095-unique-binary-search-trees-ii/0095-unique-binary-search-trees-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<TreeNode> generateTrees(int n) {
        return helper(1, n);
    }

    private List<TreeNode> helper(int low, int high){
        List<TreeNode> result = new ArrayList<>();
        if(low > high){
            result.add(null);
            return result;
        }

        for(int node=low;node<=high;node++){
            List<TreeNode> left = helper(low, node-1);
            List<TreeNode> right = helper(node+1, high);

            for(TreeNode l: left){
                for(TreeNode r: right){
                    TreeNode root = new TreeNode(node);
                    root.left = l;
                    root.right = r;
                    result.add(root);
                }
            }
        }

        return result;
    }
}